File: rede-bayseana-smile/tutorials/csharp/Tutorial9.cs
using System;
using System.Collections.Generic;
using Smile;

// Tutorial9 loads HeparII.xdsl file
// and runs multiple diagnostic algorithms
// Use the link below to download the HeparII.xdsl file:
// https://support.bayesfusion.com/docs/Examples/Discrete%20Bayesian%20Networks/HeparII.xdsl

namespace SmileNetTutorial
{
    class Tutorial9
    {
        public static void Run()
        {
            Console.WriteLine(System.IO.Directory.GetCurrentDirectory());
            Console.WriteLine("Starting Tutorial9...");
            Network net = new Network();
            net.ReadFile("HeparII.xdsl");
            Console.WriteLine("Hepar model loaded");
            PrintDiagTypes(net);

            Console.WriteLine("Creating diagnostic network object");
            DiagNetwork diag = new DiagNetwork(net);
            PrintFaultIndices(diag);

            int pursuedFaultIdx = diag.GetPursuedFault();
            Console.WriteLine("The default (most likely) pursued fault is at index {0}: {1}={2}", 
                pursuedFaultIdx,
                diag.GetFaultNodeId(pursuedFaultIdx), diag.GetFaultOutcomeId(pursuedFaultIdx));

            Console.WriteLine("Running diagnosis with no instantiated observations");
            DiagResults diagResults = diag.Update();
            PrintDiagResults(net, diagResults);

            Console.WriteLine("Running diagnosis with three observations");
            diag.InstantiateObservation("jaundice", "present");
            diag.InstantiateObservation("nausea", "absent");
            diag.InstantiateObservation("obesity", "present");
            diagResults = diag.Update();
            PrintDiagResults(net, diagResults);

            Console.WriteLine("Running diagnosis with two observations and " 
                + "focusing on Hyperbilirubinemia");
            diag.ReleaseObservation("nausea");
            diag.InstantiateObservation("obesity", "absent");
            int hyperbilirubinemiaIdx = diag.GetFaultIndex("Hyperbilirubinemia", "present");
            diag.SetPursuedFault(hyperbilirubinemiaIdx);
            diagResults = diag.Update();
            PrintDiagResults(net, diagResults);

            Console.WriteLine("Switching algorithm to cross-entropy, "
                + "observations and pursued fault unchanged");
            diag.SingleFaultAlgorithm = DiagNetwork.SingleFaultAlgorithmType.Crossentropy;
            diagResults = diag.Update();
            PrintDiagResults(net, diagResults);

            Console.WriteLine("Running diagnosis with two observations and " 
                + "focusing on both Hyperbilirubinemia and Steatosis");
            int steatosisIdx = diag.GetFaultIndex("Steatosis", "present");
            diag.SetPursuedFaults(new int[] { steatosisIdx, hyperbilirubinemiaIdx });
            diagResults = diag.Update();
            PrintDiagResults(net, diagResults);

            Console.WriteLine("Swtiching algorithm to L2 distance, "
                + "observations and pursued faults unchanged");
            diag.MultiFaultAlgorithm = DiagNetwork.MultiFaultAlgorithmType.L2NormalizedDistance;
            diagResults = diag.Update();
            PrintDiagResults(net, diagResults);

            Console.WriteLine("Tutorial9 complete.");
        }

        public static void PrintDiagTypes(Network net)
        {
            Console.WriteLine("Network has {0} nodes", net.GetNodeCount());
            Console.WriteLine("{0,-20} {1,-30} {2,-30}", 
                "Node Id", "Diagnostic Type", "Fault Outcomes");
            string[] nodeIds = net.GetAllNodeIds();
            foreach (string nodeId in nodeIds)
            {
                Network.NodeDiagType diagType = net.GetNodeDiagType(nodeId);
                List<string> faultOutcomes = new List<string>();
                if (diagType == Network.NodeDiagType.Fault)
                {
                    foreach (string outcomeId in net.GetOutcomeIds(nodeId))
                    {
                        if (net.IsFaultOutcome(nodeId, outcomeId))
                        {
                            faultOutcomes.Add(outcomeId);
                        }
                    }
                }
                Console.WriteLine("{0,-20} {1,-30} {2,-30}", nodeId, diagType,
                    (faultOutcomes.Count != 0) ? string.Join(", ", faultOutcomes) : "n/a");
            }
        }

        public static void PrintFaultIndices(DiagNetwork diag)
        {
            int faultCount = diag.FaultCount;
            Console.WriteLine("Diagnostic network has {0} faults (node/outcome pairs)", 
                faultCount);
            Console.WriteLine("{0,-12} {1,-20} {2,-20}", 
                "Fault Index", "Fault Node Id", "Fault Outcome Id");
            for (int fIdx = 0; fIdx < faultCount; fIdx++)
            {
                Console.WriteLine("{0,-12} {1,-20} {2,-20}", fIdx, 
                    diag.GetFaultNodeId(fIdx), diag.GetFaultOutcomeId(fIdx));
            }
        }

        public static void PrintFaultInfo(Network net, FaultInfo fault)
        {
            Console.WriteLine("{0,-20} {1,-20} {2,-24} {3,-8}", 
                net.GetNodeId(fault.Node), net.GetOutcomeId(fault.Node, fault.Outcome), 
                fault.Probability, fault.IsPursued ? "Yes" : "No");
        }

        public static void PrintObservationInfo(Network net, ObservationInfo observation)
        {
            Console.WriteLine("{0,-20} {1,-24} {2,-8} {3,-20}", 
                net.GetNodeId(observation.Node), observation.Measure,
                observation.Cost, observation.InfoGain);
        }

        public static void PrintDiagResults(Network net, DiagResults diagResults)
        {
            Console.WriteLine("Diag results start\n\nFaults:");
            Console.WriteLine("{0,-20} {1,-20} {2,-24} {3,-8}", 
                "Node Id", "Outcome Id", "Probability", "Is Pursued");
            foreach (FaultInfo faultInfo in diagResults.Faults)
            {
                PrintFaultInfo(net, faultInfo);
            }
            Console.WriteLine("\nObservations:");
            Console.WriteLine("{0,-20} {1,-24} {2,-8} {3,-20}", 
                "Node Id", "Measure", "Cost", "InfoGain");
            foreach (ObservationInfo observationInfo in diagResults.Observations)
            {
                PrintObservationInfo(net, observationInfo);
            }
            Console.WriteLine("\nDiag results end\n");
        }
    }
}

File: rede-bayseana-smile/tutorials/java/src/tutorials/Tutorial9.java
package tutorials;

import smile.*;

import java.util.ArrayList;
import java.util.List;

// Tutorial9 loads HeparII.xdsl file
// and runs multiple diagnostic algorithms
// Use the link below to download the HeparII.xdsl file:
// https://support.bayesfusion.com/docs/Examples/Discrete%20Bayesian%20Networks/HeparII.xdsl

public class Tutorial9 {
    public static void run() {
        System.out.println("Starting diagnosis example");
        Network net = new Network();
        net.readFile("HeparII.xdsl");
        System.out.println("Hepar model loaded");
        printDiagTypes(net);

        System.out.println("Creating diagnostic network object");
        DiagNetwork diag = new DiagNetwork(net);
        printFaultIndices(diag);

        int pursuedFaultIdx = diag.getPursuedFault();
        System.out.printf("The default (most likely) pursued fault is at index %d: %s=%s\n", 
            pursuedFaultIdx, diag.getFaultNodeId(pursuedFaultIdx), 
            diag.getFaultOutcomeId(pursuedFaultIdx));

        System.out.println("Running diagnosis with no instantiated observations");
        DiagResults diagResults = diag.update();
        printDiagResults(net, diagResults);

        System.out.println("Running diagnosis with three observations");
        diag.instantiateObservation("jaundice", "present");
        diag.instantiateObservation("nausea", "absent");
        diag.instantiateObservation("obesity", "present");
        diagResults = diag.update();
        printDiagResults(net, diagResults);

        System.out.println("Running diagnosis with two observations and "
            + "focusing on Hyperbilirubinemia");
        diag.releaseObservation("nausea");
        diag.instantiateObservation("obesity", "absent");
        int hyperbilirubinemiaIdx = diag.getFaultIndex("Hyperbilirubinemia", "present");
        diag.setPursuedFault(hyperbilirubinemiaIdx);
        diagResults = diag.update();
        printDiagResults(net, diagResults);

        System.out.println("Switching algorithm to cross-entropy, " 
            + "observations and pursued fault unchanged");
        diag.setSingleFaultAlgorithm(DiagNetwork.SingleFaultAlgorithmType.CROSSENTROPY);
        diagResults = diag.update();
        printDiagResults(net, diagResults);

        System.out.println("Running diagnosis with two observations and " 
            + "focusing on both Hyperbilirubinemia and Steatosis");
        int steatosisIdx = diag.getFaultIndex("Steatosis", "present");
        diag.setPursuedFaults(new int[]{steatosisIdx, hyperbilirubinemiaIdx});
        diagResults = diag.update();
        printDiagResults(net, diagResults);

        System.out.println("Swtiching algorithm to L2 distance, " 
            + "observations and pursued faults unchanged");
        diag.setMultiFaultAlgorithm(DiagNetwork.MultiFaultAlgorithmType.L2_NORMALIZED_DISTANCE);
        diagResults = diag.update();
        printDiagResults(net, diagResults);
                
        System.out.println("Tutorial9 complete.");
    }

    public static void printDiagTypes(Network net) {
        System.out.printf("Network has %d nodes\n", net.getNodeCount());
        System.out.printf("%-20s %-30s %-30s\n", 
            "Node Id", "Diagnostic Type", "Fault Outcomes");
        String[] nodeIds = net.getAllNodeIds();
        for (String nodeId : nodeIds) {
            int diagType = net.getNodeDiagType(nodeId);
            List<String> faultOutcomes = new ArrayList<>();
            if (diagType == Network.NodeDiagType.FAULT) {
                for (String outcomeId : net.getOutcomeIds(nodeId)) {
                    if (net.isFaultOutcome(nodeId, outcomeId)) {
                        faultOutcomes.add(outcomeId);
                    }
                }
            }
            System.out.printf("%-20s %-30s %-30s\n", nodeId, diagType,
                (faultOutcomes.size() != 0) ? String.join(", ", faultOutcomes) : "n/a");
        }
    }

    public static void printFaultIndices(DiagNetwork diag) {
        int faultCount = diag.getFaultCount();
        System.out.printf("Diagnostic network has %d faults (node/outcome pairs)\n",
            faultCount);
        System.out.printf("%-12s %-20s %-20s\n", 
            "Fault Index", "Fault Node Id", "Fault Outcome Id");
        for (int fIdx = 0; fIdx < faultCount; fIdx++) {
            System.out.printf("%-12d %-20s %-20s\n", 
                fIdx, diag.getFaultNodeId(fIdx), diag.getFaultOutcomeId(fIdx));
        }
    }

    public static void printFaultInfo(Network net, FaultInfo fault) {
        System.out.printf("%-20s %-20s %-24s %-8s\n", 
            net.getNodeId(fault.node), net.getOutcomeId(fault.node, fault.outcome), 
            fault.probability, fault.isPursued ? "Yes" : "No");
    }

    public static void printObservationInfo(Network net, ObservationInfo observation) {
        System.out.printf("%-20s %-24s %-8s %-20s\n", 
            net.getNodeId(observation.node), observation.measure,
            observation.cost, observation.infoGain);
    }

    public static void printDiagResults(Network net, DiagResults diagResults) {
        System.out.println("Diag results start\n\nFaults:");
        System.out.printf("%-20s %-20s %-24s %-8s\n", 
            "Node Id", "Outcome Id", "Probability", "Is Pursued");
        for (FaultInfo faultInfo : diagResults.faults) {
            printFaultInfo(net, faultInfo);
        }
        System.out.println("\nObservations:");
        System.out.printf("%-20s %-24s %-8s %-20s\n", 
            "Node Id", "Measure", "Cost", "InfoGain");
        for (ObservationInfo observationInfo : diagResults.observations) {
            printObservationInfo(net, observationInfo);
        }
        System.out.println("\nDiag results end\n");
    }
}
